parser.go

```diff
@@ -157,6 +157,13 @@ func (p *Parser) Progress() float32 {
 	return float32(p.currentFrame) / float32(p.header.PlaybackFrames)
 }
 
+// Weapons returns array which is used to remember what a weapon is (p250 / cz etc.).
+//
+// Note: this array is filled with meaningful values after DataTablesParsed event is dispatched.
+func (p *Parser) Weapons() [maxEntities]common.Equipment {
```

Comment:
I think we'll want to return a map here. I never really liked the fact we were using an array for weapons anyway.

Also it should be a copy of the underlying data, to avoid giving consumers the power to overwrite internal parser data.

Iirc this will take some work to adjust properly, last time I tried this it didn't go so well.

Reply:
I understand your concerns about giving a consumer the ablity to change the data, but in this particular case I actually used the fact that in Go arrays are passed by value so this actually _is_ a copy 🙂 That's the reason I [added a test to parser_test.g](https://github.com/markus-wa/demoinfocs-golang/blob/49ed3f041d7be50abf12941c376505fa6622e9e2/parser_test.go#L69)o which ensures that we cannot change the original array.

As for returning a map, yeah, this may be better for the end-user understanding and more in line with other library methods. The reason I didn't make a map is the efficiency of copying an array using  golang internals vs making a new map and filling it. It is quite a pity in Go we cannot pass a map by constant reference like in c++ -- actually being a former c++ coder it hurts my feelings to know that every time I use something like `GameState.Participants.Playing()` a full copy of the map is created 😃 So, if you really think we should return a map, let me know and I certainly will change the behaviour. I don't really think it will strike the performance of my application since it already uses quite a lot of map-copying.

TBH though I would be happy if there were more methods like [this](https://github.com/markus-wa/demoinfocs-golang/blob/be93db4b8c9cbc60cf0942b52ab455b1f29b50bf/parser.go#L102) which do not return a copy and return a map reference instead with remark about "Intended for internal usage" or something like that, so user could use them on their own head with theoretical possibility of internal data corruption but without having to copy a map every time they need access to the data.

Could you tell me a bit more about difficulties you encountered the last time so I could make some adjustments? Right now I don't really understand whan do you mean because I've already used this fork to draw a simple 2D view of a game and didn't faced any problems with it so far.

Reply:
I'm a bit suprised to see that this test passes. Slices are actually passed as a `SliceHeader` which contains a pointer to the underlying array, so it should be like passing by reference (https://stackoverflow.com/questions/39993688/are-golang-slices-passed-by-value).

And what I meant is to not only change the `Weapons()` function but also change the internal representaton from a slice to a map (`Parser.weapons`). I will try to look into this again - you can give it a go as well if you feel like it of course.

And yeah Go's immutability support is terrible, but considering the amount of CPU time spent on processing a demo file, the work to copy a map is so tiny that it's negligible imo. The benefits outweigh the negatives for me. What I wouldn't want to do is turn the slice into a map only inside `Weapons()`, in that case i'd prefer your current implementation

Reply:
I suppose we all work so often with slices in Go that sometimes we forget that such thing as "an array" does even exist 😄 

Well I don't feel like I could break something if I replace your current array-representation with a map one myself. Don't even have to replace the tese code 😃 So I think I will add the changes to the PR so you could review them later.

Reply:
awesome work man, looks good!